src/models/chainspec.rs

```diff
@@ -0,0 +1,392 @@
+use std::{collections::HashMap, time::Duration};
+
+use ethereum_types::{H160, H256, U256, U64};
+use serde::{de, Deserialize};
+
+type NodeUrl = String;
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct ChainSpec {
+    name: String,
+    data_dir: String,
+    bootnodes: Vec<NodeUrl>,
+    engine: Engine,
+    hardforks: HardForks,
+    params: Params,
+    genesis: Genesis,
+    precompiles: HashMap<String, HashMap<String, Precompiles>>,
+    balances: HashMap<H160, U256>,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct Engine {
+    name: String,
+    params: EngineParams,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct EngineParams {
+    #[serde(deserialize_with = "deserialize_period_as_duration")]
+    period: Duration,
+    #[serde(deserialize_with = "deserialize_float_as_u64")]
+    epoch: u64,
+    genesis: EngineGenesis,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct EngineGenesis {
+    vanity: H256,
+    signers: Vec<H160>,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct HardForks {
+    eip140: U64,
```

Comment:
Here and below: is it possible to keep standard Rust `u64`? You could try `deserialize_with` to parse into `primitive_types::U64` and from there extract `u64`.